call_option.go

```diff
@@ -1,118 +1,97 @@
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
 package gax
 
 import (
+	"math/rand"
 	"time"
 
+	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
 )
 
 type CallOption interface {
 	Resolve(*CallSettings)
 }
 
-type callOptions []CallOption
-
-func (opts callOptions) Resolve(s *CallSettings) *CallSettings {
-	for _, opt := range opts {
-		opt.Resolve(s)
-	}
-	return s
-}
-
-// Encapsulates the call settings for a particular API call.
-type CallSettings struct {
-	Timeout       time.Duration
-	RetrySettings RetrySettings
-}
+type retryFuncOption func() RetryFunc
 
-// Per-call configurable settings for retrying upon transient failure.
-type RetrySettings struct {
-	RetryCodes      map[codes.Code]bool
-	BackoffSettings BackoffSettings
+func (o retryFuncOption) Resolve(s *CallSettings) {
+	s.Retry = o
 }
 
-// Parameters to the exponential backoff algorithm for retrying.
-type BackoffSettings struct {
-	DelayTimeoutSettings MultipliableDuration
-	RPCTimeoutSettings   MultipliableDuration
+func WithRetry(fn func() RetryFunc) CallOption {
+	return retryFuncOption(fn)
 }
 
-type MultipliableDuration struct {
-	Initial    time.Duration
-	Max        time.Duration
-	Multiplier float64
+type Backoff struct {
+	Initial time.Duration
+	Max     time.Duration
+	Mult    float64
+	Codes   []codes.Code
 }
 
-func (w CallSettings) Resolve(s *CallSettings) {
-	s.Timeout = w.Timeout
-	s.RetrySettings = w.RetrySettings
-
-	s.RetrySettings.RetryCodes = make(map[codes.Code]bool, len(w.RetrySettings.RetryCodes))
-	for key, value := range w.RetrySettings.RetryCodes {
-		s.RetrySettings.RetryCodes[key] = value
+func (r *Backoff) Retry(err error) (time.Duration, bool) {
+	c := grpc.Code(err)
+	for _, rc := range r.Codes {
+		if c == rc {
+			return r.pause(), true
+		}
 	}
+	return 0, false
 }
 
-type withTimeout time.Duration
-
-func (w withTimeout) Resolve(s *CallSettings) {
-	s.Timeout = time.Duration(w)
-}
-
-// WithTimeout sets the client-side timeout for API calls if the call isn't
-// retrying.
-func WithTimeout(timeout time.Duration) CallOption {
-	return withTimeout(timeout)
-}
-
-type withRetryCodes []codes.Code
-
-func (w withRetryCodes) Resolve(s *CallSettings) {
-	s.RetrySettings.RetryCodes = make(map[codes.Code]bool)
-	for _, code := range []codes.Code(w) {
-		s.RetrySettings.RetryCodes[code] = true
+func (r *Backoff) pause() time.Duration {
+	if r.Initial == 0 {
+		r.Initial = time.Second
 	}
+	if r.Max == 0 {
+		r.Max = 30 * time.Second
+	}
+	if r.Mult == 0 {
+		r.Mult = 2
+	}
```

Comment:
It's strange that this function mutates r.

I'd not expect that by looking at its function signature.